File: mobile/lib/core/api/meetings_api.dart
import 'package:dio/dio.dart';
import '../../models/meeting.dart';
import '../../models/transcription.dart';
import 'api_client.dart';

class MeetingsApi {
  final _dio = ApiClient.instance.dio;

  Future<List<Meeting>> list() async {
    final res = await _dio.get('/meetings');
    return (res.data['data'] as List).map((e) => Meeting.fromJson(e)).toList();
  }

  Future<Meeting> create({
    required int projectId,
    required String title,
    required DateTime scheduledAt,
    String? location,
  }) async {
    final res = await _dio.post('/meetings', data: {
      'projectId': projectId,
      'title': title,
      'scheduledAt': scheduledAt.toIso8601String(),
      if (location != null) 'location': location,
    });
    return Meeting.fromJson(res.data);
  }

  Future<Meeting> findOne(int id) async {
    final res = await _dio.get('/meetings/$id');
    return Meeting.fromJson(res.data);
  }

  Future<Map<String, dynamic>> uploadAudio({
    required int meetingId,
    required String filePath,
    int? durationSeconds,
    void Function(int sent, int total)? onSendProgress,
  }) async {
    final form = FormData.fromMap({
      'file': await MultipartFile.fromFile(filePath),
      'source': 'MOBILE',
      if (durationSeconds != null) 'durationSeconds': durationSeconds,
    });
    final res = await _dio.post(
      '/meetings/$meetingId/audio',
      data: form,
      onSendProgress: onSendProgress,
      options: Options(
        sendTimeout: const Duration(minutes: 30),
        receiveTimeout: const Duration(minutes: 5),
      ),
    );
    return Map<String, dynamic>.from(res.data);
  }

  Future<Transcription?> getTranscription(int meetingId) async {
    final res = await _dio.get('/meetings/$meetingId/transcription');
    if (res.data == null) return null;
    return Transcription.fromJson(res.data);
  }
}

File: web/src/api/meetings.api.ts
import { api } from './client';
import type {
  Acta,
  AudioFile,
  BacklogResult,
  Meeting,
  MeetingType,
  Paginated,
  Transcription,
} from './types';

export const meetingsApi = {
  listByProject: (projectId: number) =>
    api.get<Paginated<Meeting>>('/meetings', { params: { projectId } }).then((r) => r.data),
  create: (body: {
    projectId: number;
    title: string;
    description?: string;
    scheduledAt: string;
    location?: string;
    meetingType?: MeetingType;
  }) => api.post<Meeting>('/meetings', body).then((r) => r.data),
  findOne: (id: number) => api.get<Meeting>(`/meetings/${id}`).then((r) => r.data),
  uploadAudio: (meetingId: number, file: File) => {
    const form = new FormData();
    form.append('file', file);
    form.append('source', 'WEB');
    return api
      .post(`/meetings/${meetingId}/audio`, form, {
        headers: { 'Content-Type': 'multipart/form-data' },
      })
      .then((r) => r.data);
  },
  getTranscription: (meetingId: number) =>
    api.get<Transcription | null>(`/meetings/${meetingId}/transcription`).then((r) => r.data),
  getAudio: (meetingId: number) =>
    api.get<AudioFile | null>(`/meetings/${meetingId}/audio`).then((r) => r.data),
  getAudioBlobUrl: async (audioId: number) => {
    const res = await api.get(`/audio/${audioId}/stream`, { responseType: 'blob' });
    const blob = new Blob([res.data], { type: res.headers['content-type'] || 'audio/webm' });
    return URL.createObjectURL(blob);
  },
  generateActa: (meetingId: number) =>
    api.post<Acta>(`/meetings/${meetingId}/acta/generate`).then((r) => r.data),
  getActa: (meetingId: number) =>
    api.get<Acta | null>(`/meetings/${meetingId}/acta`).then((r) => r.data),
  update: (id: number, body: Partial<{ title: string; description: string; scheduledAt: string; location: string; status: Meeting['status'] }>) =>
    api.patch<Meeting>(`/meetings/${id}`, body).then((r) => r.data),
  close: (id: number) =>
    api.patch<Meeting>(`/meetings/${id}`, { status: 'CLOSED' }).then((r) => r.data),
  reopen: (id: number) =>
    api.patch<Meeting>(`/meetings/${id}`, { status: 'SCHEDULED' }).then((r) => r.data),
};

export const actasApi = {
  findOne: (id: number) => api.get<Acta>(`/actas/${id}`).then((r) => r.data),
  update: (id: number, contentMarkdown: string) =>
    api.patch<Acta>(`/actas/${id}`, { contentMarkdown }).then((r) => r.data),
  submitReview: (id: number) => api.post<Acta>(`/actas/${id}/submit-review`).then((r) => r.data),
  approve: (id: number) => api.post<Acta>(`/actas/${id}/approve`).then((r) => r.data),
  regenerate: (id: number) => api.post<Acta>(`/actas/${id}/regenerate`).then((r) => r.data),
  generateBacklog: (id: number) =>
    api.post<BacklogResult>(`/actas/${id}/backlog`).then((r) => r.data),
  downloadPdf: async (id: number) => {
    const res = await api.get(`/actas/${id}/pdf`, { responseType: 'blob' });
    const blob = new Blob([res.data], { type: 'application/pdf' });
    const url = URL.createObjectURL(blob);
    const a = document.createElement('a');
    a.href = url;
    a.download = `acta-${id}.pdf`;
    document.body.appendChild(a);
    a.click();
    a.remove();
    setTimeout(() => URL.revokeObjectURL(url), 1000);
  },
};
